discovery-provider/src/challenges/challenge.py

```diff
@@ -44,99 +78,173 @@ class ChallengeManager:
                 )
             )
 
         if not event_metadatas:
             return
 
-        user_ids = list(map(lambda x: x["user_id"], event_metadatas))
-        user_id_metadatas = {x["user_id"]: x for x in event_metadatas}
+        # Add specifiers
+        events_with_specifiers: List[FullEventMetadata] = [
+            {
+                "user_id": event["user_id"],
+                "block_number": event["block_number"],
+                "extra": event["extra"],
+                "specifier": self._updater.generate_specifier(
+                    event["user_id"], event["extra"]
+                ),
+            }
+            for event in event_metadatas
+        ]
+
+        # Drop any duplicate specifiers
+        events_with_specifiers_map = {
+            event["specifier"]: event for event in events_with_specifiers
+        }
+        events_with_specifiers = list(events_with_specifiers_map.values())
+
+        specifiers: List[str] = [e["specifier"] for e in events_with_specifiers]
 
         # Gets all user challenges,
         existing_user_challenges = fetch_user_challenges(
-            session, self.challenge_id, user_ids
+            session, self.challenge_id, specifiers
         )
 
         # Create users that need challenges still
-        existing_user_ids = {
-            challenge.user_id for challenge in existing_user_challenges
+        existing_specifiers = {
+            challenge.specifier for challenge in existing_user_challenges
         }
-        needs_challenge_ids = list(
-            {id for id in user_ids if not id in existing_user_ids}
-        )
+
+        # Create new challenges
+
+        new_challenge_metadata = [
+            metadata
+            for metadata in events_with_specifiers
+            if metadata["specifier"] not in existing_specifiers
+        ]
+        to_create_metadata: List[FullEventMetadata] = []
+        if self._challenge_type == ChallengeType.aggregate:
+            # For aggregate challenges, only create them
+            # if we haven't maxed out completion yet, and
+            # we haven't overriden this via should_create_new_challenge
+
+            # Get *all* UserChallenges per user
+            user_ids = list({e["user_id"] for e in event_metadatas})
+            all_user_challenges: List[Tuple[int, int]] = (
+                session.query(
+                    UserChallenge.user_id, func.count(UserChallenge.specifier)
+                )
+                .filter(
+                    UserChallenge.challenge_id == self.challenge_id,
+                    UserChallenge.user_id.in_(user_ids),
+                )
+                .group_by(UserChallenge.user_id)
+            ).all()
+            challenges_per_user = dict(all_user_challenges)
+            for new_metadata in new_challenge_metadata:
+                completion_count = challenges_per_user.get(new_metadata["user_id"], 0)
+                if self._step_count and completion_count >= self._step_count:
```

Comment:
Will aggregate challenges not always have a step_count?

Reply:
theoretically could be an unlimited aggregate challenge